paths_cli/tests/compiling/test_core.py

```diff
@@ -0,0 +1,319 @@
+        assert builder(string="foo") == "oofoo"
+
+class TestInstanceCompilerPlugin:
+    @staticmethod
+    def _builder(req_param, opt_default=10, opt_override=100):
+        return f"{req_param}, {opt_default}, {opt_override}"
+
+    def setup(self):
+        identity = lambda x: x
+        self.parameters = [
+            Parameter('req_param', identity, json_type="string"),
+            Parameter('opt_default', identity, json_type="int", default=10),
+            Parameter('opt_override', identity, json_type='int',
+                      default=100)
+        ]
+        self.instance_builder = InstanceCompilerPlugin(
+            self._builder,
+            self.parameters,
+            name='demo',
+            aliases=['foo', 'bar'],
+        )
+        self.instance_builder.category = 'demo'
+        self.input_dict = {'req_param': "qux", 'opt_override': 25}
+
+    def test_schema_name(self):
+        assert self.instance_builder.schema_name == 'demo'
+        self.instance_builder.category = 'foo'
+        assert self.instance_builder.schema_name == 'demo-foo'
+        self.instance_builder.name = 'demo-foo'
+        assert self.instance_builder.schema_name == 'demo-foo'
+
+
+    def test_to_json_schema(self):
+        # to_json_schema should create a valid JSON schema entry for this
+        # instance builder
+        # TODO: this may change as I better understand JSON schema. Details
+        # of the JSON schema API aren't locked until I can build docs from
+        # our schema.
+        expected_schema = {
+            'properties': {
+                'name': {'type': 'string'},
+                'type': {'type': 'string',
+                         'enum': ['demo']},
+                'req_param': {'type': 'string', 'description': None},
+                'opt_default': {'type': 'int', 'description': None},
+                'opt_override': {'type': 'int', 'description': None},
+            },
+            'required': ['req_param'],
+        }
+        name, schema =  self.instance_builder.to_json_schema()
+        assert name == 'demo'
+        assert expected_schema['required'] == schema['required']
+        assert expected_schema['properties'] == schema['properties']
+
+    def test_compile_attrs(self):
+        # compile_attrs should create a dictionary with correct objects in
+        # the attributes from the input dictionary
+        expected = {'req_param': "qux", 'opt_override': 25}
+        # note that the parameter where we use the default value isn't
+        # listed: the default value should match the default used in the
+        # code, though!
+        compile_attrs = self.instance_builder.compile_attrs
+        assert compile_attrs(self.input_dict) == expected
+
+    def test_compile_attrs_compiler_integration(self):
+        # compile_attrs gives the same object as already existing in a
+        # compiler if one of the parameters uses that compiler to load a
+        # named object
+        pytest.skip()
```

Comment:
is this one forgotten or consciously not implemented?

Reply:
Implemented skipped tests